src/LanguageServer/Impl/Indexing/IndexParser.cs

```diff
@@ -37,34 +41,64 @@ internal sealed class IndexParser : IIndexParser {
             _version = version;
         }
 
-        public void Dispose() {
-            _allProcessingCts.Cancel();
-            _allProcessingCts.Dispose();
+        public Task ParseAsync(string path, CancellationToken cancellationToken = default) {
+            lock (_syncObj) {
+                CancelCurrentParse();
+                _tcs = new TaskCompletionSource<object>();
+                _linkedParseCts = CancellationTokenSource.CreateLinkedTokenSource(_allProcessingCts.Token, cancellationToken);
+                _parseTask = Task.Run(() => Parse(path, _linkedParseCts));
+                return _tcs.Task;
+            }
         }
 
-        public Task<bool> ParseAsync(string path, CancellationToken parseCancellationToken = default) {
-            var linkedParseCts = CancellationTokenSource.CreateLinkedTokenSource(_allProcessingCts.Token, parseCancellationToken);
-            var linkedParseToken = linkedParseCts.Token;
-            return Task<bool>.Run((System.Func<bool>)(() => {
-                if (!_fileSystem.FileExists(path)) {
-                    return false;
-                }
+        private void CancelCurrentParse() {
+            _linkedParseCts?.Cancel();
+            _linkedParseCts?.Dispose();
+            _linkedParseCts = null;
+            _parseTask = null;
+        }
+
+        private void Parse(string path, CancellationTokenSource parseCts) {
+            if (parseCts.Token.IsCancellationRequested) {
+                _tcs.SetCanceled();
+                parseCts.Token.ThrowIfCancellationRequested();
+            }
+            if (_fileSystem.FileExists(path)) {
                 try {
-                    linkedParseToken.ThrowIfCancellationRequested();
+                    if (parseCts.Token.IsCancellationRequested) {
+                        _tcs.SetCanceled();
+                        parseCts.Token.ThrowIfCancellationRequested();
+                    }
                     using (var stream = _fileSystem.FileOpen(path, FileMode.Open)) {
                         var parser = Parser.CreateParser(stream, _version);
-                        linkedParseToken.ThrowIfCancellationRequested();
                         _symbolIndex.Add(path, parser.ParseFile());
-                        return true;
                     }
                 } catch (FileNotFoundException e) {
                     Trace.TraceError(e.Message);
-                    return false;
                 }
-            })).ContinueWith((Task<bool> task) => {
-                linkedParseCts.Dispose();
-                return task.Result;
-            }, linkedParseToken);
+            }
+            lock (_syncObj) {
+                if (_linkedParseCts == parseCts) {
+                    _linkedParseCts.Dispose();
+                    _linkedParseCts = null;
+                    _tcs.SetResult(new object());
```

Comment:
See my other comment. This will never get called if code above throws cancellation.